src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -638,6 +659,132 @@ await OpenCore(
+            if (statusCode != NegotiateAuthenticationStatusCode.ContinueNeeded)
+            {
+                // Unable to retrieve credentials
+                // If it's required, throw an appropriate exception
+                if (isRequired)
+                    throw new NpgsqlException($"Unable to negotiate GSS encryption: {statusCode}");
+
+                return GssEncryptionResult.GetCredentialFailure;
+            }
+
+            WriteGSSEncryptRequest();
+            await Flush(async, cancellationToken).ConfigureAwait(false);
+
+            await ReadBuffer.Ensure(1, async).ConfigureAwait(false);
+            var response = (char)ReadBuffer.ReadByte();
+
+            // TODO: Server can respond with an error here
+            // but according to documentation we shouldn't display this error to the user/application
+            // since the server has not been authenticated (CVE-2024-10977)
+            // See https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-GSSAPI
+            switch (response)
+            {
+            default:
+                throw new NpgsqlException($"Received unknown response {response} for GSSEncRequest (expecting G or N)");
+            case 'N':
+                return GssEncryptionResult.NegotiateFailure;
+            case 'G':
+                break;
+            }
+
+            if (ReadBuffer.ReadBytesLeft > 0)
+                throw new NpgsqlException(
+                    "Additional unencrypted data received after GSS encryption negotiation - this should never happen, and may be an indication of a man-in-the-middle attack.");
+
+            var lengthBuffer = new byte[4];
+
+            await WriteGssEncryptMessage(async, data, lengthBuffer).ConfigureAwait(false);
+
+            while (true)
+            {
+                if (async)
+                    await _stream.ReadExactlyAsync(lengthBuffer, cancellationToken).ConfigureAwait(false);
+                else
+                    _stream.ReadExactly(lengthBuffer);
+
+                var messageLength = BitConverter.IsLittleEndian
+                    ? BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<int>(ref lengthBuffer[0]))
+                    : Unsafe.ReadUnaligned<int>(ref lengthBuffer[0]);
+
+                var buffer = ArrayPool<byte>.Shared.Rent(messageLength);
+                if (async)
+                    await _stream.ReadExactlyAsync(buffer.AsMemory(0, messageLength), cancellationToken).ConfigureAwait(false);
+                else
+                    _stream.ReadExactly(buffer.AsSpan(0, messageLength));
+
+                data = authentication.GetOutgoingBlob(buffer.AsSpan(0, messageLength), out statusCode);
+                ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
+                if (statusCode is not NegotiateAuthenticationStatusCode.Completed and not NegotiateAuthenticationStatusCode.ContinueNeeded)
```

Comment:
nit: maybe just switch on the statusCode, breaking if it's Completed (rather than checking `data is null` below)?

Reply:
The problem is, it can return `NegotiateAuthenticationStatusCode.Completed` but still has some data for us to send over (at least we had that issue with gss/sspi auth). While I can't confirm that it holds true for gss encryption, it's probably better to have similar logic?